src/background_thread.c

```diff
@@ -0,0 +1,554 @@
+		}
+		info->started = false;
+		nstime_init(&info->next_wakeup, 0);
+		info->npages_to_purge_new = 0;
+	}
+
+	return false;
+}
+
+static inline bool
+set_current_thread_affinity(UNUSED int cpu) {
+#if defined(JEMALLOC_HAVE_SCHED_SETAFFINITY)
+	cpu_set_t cpuset;
+	CPU_ZERO(&cpuset);
+	CPU_SET(cpu, &cpuset);
+	int ret = sched_setaffinity(0, sizeof(cpu_set_t), &cpuset);
+
+	return (ret != 0);
+#else
+	return false;
+#endif
+}
+
+/* Threshold for determining when to wake up the background thread. */
+#define BACKGROUND_THREAD_NPAGES_THRESHOLD UINT64_C(1024)
+#define BILLION UINT64_C(1000000000)
+/* Minimal sleep interval 100 ms. */
+#define BACKGROUND_THREAD_MIN_INTERVAL_NS (BILLION / 10)
+#define BACKGROUND_THREAD_INDEFINITE_SLEEP UINT64_MAX
+
+static void
+decay_purge_interval_search(arena_decay_t *decay, size_t lb, size_t ub,
+    size_t *next_lb, size_t *next_ub) {
+	assert(ub > lb);
+	size_t target = (ub + lb) / 2;
+
+	size_t i;
+	uint64_t sum = 0;
+	for (i = 0; i < target; i++) {
+		sum += decay->backlog[i] * h_steps[i];
+	}
+	for (; i < SMOOTHSTEP_NSTEPS; i++) {
+		sum += decay->backlog[i] * (h_steps[i] - h_steps[i - target]);
+	}
+	sum >>= SMOOTHSTEP_BFP;
+
+	if (sum > BACKGROUND_THREAD_NPAGES_THRESHOLD) {
+		*next_lb = lb;
+		*next_ub = target;
+	} else {
+		*next_lb = target;
+		*next_ub = ub;
+	}
+	assert(*next_lb <= *next_ub);
+}
+
+static uint64_t
+arena_decay_compute_purge_interval_impl(tsdn_t *tsdn, arena_decay_t *decay,
+    extents_t *extents) {
+	if (malloc_mutex_trylock(tsdn, &decay->mtx)) {
+		/* Use minimal interval if decay is contended. */
+		return BACKGROUND_THREAD_MIN_INTERVAL_NS;
+	}
+
+	uint64_t interval;
+	ssize_t decay_time = atomic_load_zd(&decay->time_ms, ATOMIC_RELAXED);
+	if (decay_time <= 0) {
+		/* Purging is eagerly done or disabled currently. */
+		interval = BACKGROUND_THREAD_INDEFINITE_SLEEP;
+		goto label_done;
+	}
+
+	uint64_t decay_interval_ns = nstime_ns(&decay->interval);
+	assert(decay_interval_ns > 0);
+	size_t npages = extents_npages_get(extents);
+	if (npages == 0) {
+		unsigned i;
+		for (i = 0; i < SMOOTHSTEP_NSTEPS; i++) {
+			if (decay->backlog[i] > 0) {
+				break;
+			}
+		}
+		if (i == SMOOTHSTEP_NSTEPS) {
+			/* No dirty pages recorded.  Sleep indefinitely. */
+			interval = BACKGROUND_THREAD_INDEFINITE_SLEEP;
+			goto label_done;
+		}
+	}
+	if (npages <= BACKGROUND_THREAD_NPAGES_THRESHOLD) {
+		/* Use max interval. */
+		interval = decay_interval_ns * SMOOTHSTEP_NSTEPS;
+		interval = interval < BACKGROUND_THREAD_MIN_INTERVAL_NS ?
+		    BACKGROUND_THREAD_MIN_INTERVAL_NS : interval;
+		goto label_done;
+	}
+
+	size_t lb = BACKGROUND_THREAD_MIN_INTERVAL_NS / decay_interval_ns;
+	/* Decay deadline offset is up to 2 decay intervals. */
+	lb = lb < 2 ? 2 : lb;
+	size_t ub = SMOOTHSTEP_NSTEPS;
+	size_t initial_lb = lb;
+	size_t initial_ub = ub;
+	/* With the current setting, bisect 5 times reaches ~6 intervals. */
+	unsigned n_search = 0;
+	size_t next_lb, next_ub;
+	while (ub > lb + 7) {
```

Comment:
What is the intent of the magic number 7 here?  If the intent is to terminate the binary search early to reduce computation, I'd argue that the termination condition should be based on how many elements fall within an acceptable time range, which differs according to both ```SMOOTHSTEP_NSTEPS``` and ```decay_time```.